tests/query_params_test.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tests
+
+import (
+	"bytes"
+	"context"
+	"go-safeweb/testing/requesttesting"
+	"net/http"
+	"testing"
+)
+
+const statusOK = "HTTP/1.1 200 OK\r\n"
+const statusBadReq = "HTTP/1.1 400 Bad Request\r\n"
+
+// isOkReq is a helper function that checks whether the server response is 200
+func isOkReq(resp []byte) bool {
+	if bytes.HasPrefix(resp, []byte(statusOK)) {
+		return true
+	}
+	return false
+}
+
+// isBadReq is a helper function that checks whether the server response is 400
+func isBadReq(resp []byte) bool {
+	if bytes.HasPrefix(resp, []byte(statusBadReq)) {
+		return true
+	}
+	return false
+}
+
+// Ensures Query() only returns a map of size one and verifies whether sending a
+// request with two values for the same key returns a []string of length 2
+// containing the correct values
+func TestMultipleQueryParametersSameKey(t *testing.T) {
+	var (
+		valueOne = "potatO"
+		valueTwo = "Tomato"
+		req      = []byte("GET /?vegetable=" + valueOne + "&vegetable=" + valueTwo + " HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "\r\n")
+	)
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		queryParams := req.URL.Query()
+		if len(queryParams) != 1 {
+			t.Errorf("len(queryParams): got %d, want %d", len(queryParams), 1)
+		}
+
+		vegetableParamValues := queryParams["vegetable"]
+		if len(vegetableParamValues) != 2 {
+			t.Errorf("len(vegetableQueryParams): got %d, want %d", len(vegetableParamValues), 2)
+		}
+		if vegetableParamValues[0] != valueOne || vegetableParamValues[1] != valueTwo {
+			t.Errorf("queryParams values: expected "+valueOne+" and "+valueTwo+", got %s and %s", vegetableParamValues[0], vegetableParamValues[1])
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
```

Comment:
In general setup functions or functions that are not expected to fail should use t.Fatal (which will abort the test).

On the other hand, when doing several assertions on a data structure, or if it in general makes sense to keep the test going, then it's recommended to use t.Error, so a developer breaking the test does not have to do several iterations to find all the errors.

Reply:
Thanks for this!